libavutil/hwcontext_vaapi.c

```diff
@@ -663,17 +665,92 @@ static int vaapi_frames_init(AVHWFramesContext *hwfc)
     return err;
 }
 
+static int vaapi_hw_sub_frames_init(AVBufferRef *ref)
+{
+    AVHWFramesContext *ctx = (AVHWFramesContext*)ref->data;
+    AVFrame **frames;
+    int i, ret = 0;
+
+
+    ret = vaapi_hw_frames_init(ctx);
+    if (ret < 0)
+        return ret;
+
+    if (ctx->internal->pool_internal && !ctx->pool)
+        ctx->pool = ctx->internal->pool_internal;
+
+    frames = av_calloc(ctx->initial_pool_size, sizeof(*frames));
+    if (!frames)
+        return AVERROR(ENOMEM);
+
+    for (i = 0; i < ctx->initial_pool_size; i++) {
+        frames[i] = av_frame_alloc();
+        if (!frames[i])
+            goto fail;
+
+        ret = av_hwframe_get_buffer(ref, frames[i], 0);
+        if (ret < 0)
+            goto fail;
+    }
+
+fail:
+    for (i = 0; i < ctx->initial_pool_size; i++) {
+        av_frame_free(&frames[i]);
+    }
+    av_freep(&frames);
+
+    return ret;
+}
+
+static int vaapi_frames_init(AVHWFramesContext *hwfc)
+{
+    VAAPIFramesContext *ctx = hwfc->internal->priv;
+    AVVAAPIFramesContext *avfc = hwfc->hwctx;
+    AVHWFramesContext *sub_frames_ctx;
+    int ret;
+
+    ret = vaapi_hw_frames_init(hwfc);
+    if (ret < 0)
+        return ret;
+
+    if (avfc->enable_sub_frame){
+        ctx->sub_frames_ref = av_hwframe_ctx_alloc(hwfc->device_ref);
+        if (!ctx->sub_frames_ref) {
+            return AVERROR(ENOMEM);
+        }
+        sub_frames_ctx = (AVHWFramesContext*)ctx->sub_frames_ref->data;
+
+        sub_frames_ctx->width             = avfc->sub_frame_width;
+        sub_frames_ctx->height            = avfc->sub_frame_height;
+        sub_frames_ctx->format            = AV_PIX_FMT_VAAPI;
+        sub_frames_ctx->sw_format         = avfc->sub_frame_sw_format;
+        sub_frames_ctx->initial_pool_size = hwfc->initial_pool_size;
+
+        ret = vaapi_hw_sub_frames_init(ctx->sub_frames_ref);
```

Comment:
Why not call av_hwframe_ctx_init() and we don't need vaapi_hw_sub_frames_init() because we can reuse vaapi_hw_frames_init(). Or is there any difference I miss?